src/connector/src/source/filesystem/opendal_source/opendal_reader.rs

```diff
@@ -0,0 +1,172 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::{Ok, Result};
+use async_trait::async_trait;
+use futures::TryStreamExt;
+use futures_async_stream::try_stream;
+use opendal::Operator;
+use risingwave_common::error::RwError;
+use tokio::io::BufReader;
+use tokio_util::io::{ReaderStream, StreamReader};
+
+use super::opendal_enumerator::OpendalEnumerator;
+use super::OpenDalSourceProperties;
+use crate::parser::{ByteStreamSourceParserImpl, ParserConfig};
+use crate::source::filesystem::{nd_streaming, OpendalFsSplit};
+use crate::source::{
+    BoxSourceWithStateStream, Column, SourceContextRef, SourceMessage, SourceMeta, SplitMetaData,
+    SplitReader, StreamChunkWithState,
+};
+
+const MAX_CHANNEL_BUFFER_SIZE: usize = 2048;
+const STREAM_READER_CAPACITY: usize = 4096;
+#[derive(Debug, Clone)]
+pub struct OpendalReader<C: OpenDalSourceProperties> {
+    connector: OpendalEnumerator<C>,
+    splits: Vec<OpendalFsSplit<C>>,
+    parser_config: ParserConfig,
+    source_ctx: SourceContextRef,
+}
+#[async_trait]
+impl<C: OpenDalSourceProperties> SplitReader for OpendalReader<C>
+where
+    C: Send + Clone + PartialEq + 'static + Sync,
+{
+    type Properties = C;
+    type Split = OpendalFsSplit<C>;
+
+    async fn new(
+        properties: C,
+        splits: Vec<OpendalFsSplit<C>>,
+        parser_config: ParserConfig,
+        source_ctx: SourceContextRef,
+        _columns: Option<Vec<Column>>,
+    ) -> Result<Self> {
+        let connector = Self::Properties::new_enumerator(properties)?;
+        let opendal_reader = OpendalReader {
+            connector,
+            splits,
+            parser_config,
+            source_ctx,
+        };
+        Ok(opendal_reader)
+    }
+
+    fn into_stream(self) -> BoxSourceWithStateStream {
+        self.into_chunk_stream()
+    }
+}
+
+impl<C: OpenDalSourceProperties> OpendalReader<C>
+where
+    C: Send + Clone + Sized + PartialEq + 'static,
+{
+    #[try_stream(boxed, ok = StreamChunkWithState, error = RwError)]
+    async fn into_chunk_stream(self) {
+        for split in self.splits {
+            let actor_id = self.source_ctx.source_info.actor_id.to_string();
+            let source_id = self.source_ctx.source_info.source_id.to_string();
+            let source_ctx = self.source_ctx.clone();
+
+            let split_id = split.id();
+
+            let data_stream =
+                Self::stream_read_object(self.connector.op.clone(), split, self.source_ctx.clone());
+
+            let parser =
+                ByteStreamSourceParserImpl::create(self.parser_config.clone(), source_ctx).await?;
+            let msg_stream = if matches!(
+                parser,
+                ByteStreamSourceParserImpl::Json(_) | ByteStreamSourceParserImpl::Csv(_)
+            ) {
+                parser.into_stream(nd_streaming::split_stream(data_stream))
+            } else {
+                parser.into_stream(data_stream)
+            };
```

Comment:
Should we ban FORMAT other than `Plain`, `Json` and `Csv` for file source？

Reply:
we've done that in the frontend 
https://github.com/risingwavelabs/risingwave/blob/1c453f43463a0c1813c38d3c9435258da4efb2f3/src/frontend/src/handler/create_source.rs#L897-L902